### PR TITLE
chore: refactor pagination

### PR DESCRIPTION
**Pull request by presleyp**
- extract pagination logic into util functions and a pagination machine
- spawn pagination machine inside the workspaces machine and the audit page machine, and pass the reference to it down to the pagination widget. 
- extract `PageButton` component from `PaginationWidget` just for clarity.

bonus features:
- call the workspaces count endpoint less often for better performance. While the workspaces are updated every 5 seconds, the count is only updated on filter.
- fix the audit log bug where filtering doesn't reset the page and you can get an impossible (and empty) page.

Workspaces page with 500 workspaces
https://user-images.githubusercontent.com/1290996/197893688-4ff7df74-81d6-4717-ace4-95fb32a44ade.mp4

Audit page with 45 pages of logs
https://user-images.githubusercontent.com/1290996/197894060-9e438822-bedf-46bc-a4ed-f96df1107db3.mp4

still to do:
- maybe separate the audit log fetch from the audit log count API call, and call count less often, like in workspaces. Depends on whether Colin plans to change the way we query count.

Things I tried and rejected:
- using `useEffect` to just sync with the URL, using the URL as source of truth. Alas, `useEffect` gets called too much.
- spinning up the pagination machine inside the pagination component so we don't have to pass a ref around: this is fine for storybook I believe, and I could pass the workspace machine id down through props so the pagination machine could send messages to it, but I couldn't find a good way for the workspace machine to be able to send messages to the pagination machine. You can set a machine's id when you `spawn` it but not through `useMachine`, apparently. Rather than having a whole "this is my name" handshake at the beginning (when I want things to load fast), I went with spawning. 
- having the pagination machine be in charge of updating the URL: much to my chagrin, `setSearchParameters` overrides all search parameters rather than merging with them. So whenever it's called, you need to know all of the current parameters; the pagination machine only knows the page. So I thought whatever page component is being paginated should manage its URL and any query parameters it uses.

### CODE REVIEW

File: site/src/components/PaginationWidget/PageButton.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      disabled={disabled}
+    >
+      <div>{page ?? placeholder}</div>
+    </Button>
+  )
+}
+
+const useStyles = makeStyles((theme) => ({
+  pageButton: {
+    "&:not(:last-of-type)": {
```
Comment (BrunoQuaresma) [inline site/src/components/PaginationWidget/PageButton.tsx:43]: TIL about `:last-of-type`

**Review by BrunoQuaresma — APPROVED** [review]
- Code looks good to me ✅ 
- I made some QA and things are working nicely ✅ 
- I understand the motivation to move that into a machine ✅ 

Wondering if at some point we could have an `apiFetchMachine` that includes fetching + error handling + pagination.

File: site/src/components/PaginationWidget/PageButton.tsx (hunk 1 above)
Comment (presleyp) [reply site/src/components/PaginationWidget/PageButton.tsx:43]: I think that was @Kira-Pilot! Just moving this into its own file.

**Comment by Kira-Pilot** [comment]
driving by to say this is cool!

**Pull request merged**